## Intermediate Data Science

#### University of Redlands - DATA 201
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data201.joannabieri.com](https://joannabieri.com/data201_intermediate.html)

## Computer Set Up

In [2]:
!python --version

Python 3.13.5


In [1]:
!conda --version

conda 25.7.0


In [3]:
!git --version

git version 2.45.2


**Clone the Repo for our Class**

<a href="https://github.com/Redlands-DATA201/FALL25" target="_blank">Redlands-DATA201/FALL2025</a>

In [4]:
# # Depending on your setup you might need to install modules
# !conda install -y numpy
# !conda install -y pandas
# !conda install -y matplotlib
# !conda install -y plotly

In [21]:
# Some basic package imports
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

# Review of DATA 101

For our review we are going to jump in and do some Exploratory Data Analysis (EDA) on a data set that we have seen before. This time you are just given a .csv file and your goal is to analyze it - answering these questions.

**I expect this to feel impossible at first!!!** but if you all work together, share code, and look things up as needed, I know you can do it!

#### This was a Homework Assignment in Data 101 - but see how much you can figure out on your own!

* Answer the following questions using reproducible Python code.
    - What does it mean to be reproducible? This means someone else at any time in the future can run and understand your code almost like reading a blog post.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

0.  Read in the .csv file using Pandas and display it

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego?

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. Then using a for loop calculate the average money spent per customer for each area code.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

**This exersize comes from: https://datasciencebox.org/course-materials/hw-instructions/hw-05/hw-05-legos**

In [22]:
file_name = 'data/lego-sales.csv'

In [23]:
DF= pd.read_csv(file_name)
DF

#pd is short for PANDAS-a python library used
#for data analysis and manipulation
#read_csv is a function in pandas that loads data 
#from a .csv file (Comma Separated Values) into a DataFrame.

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

In this dataset, each row is an observation of a Lego set purchase by a customer. The variables include customer info (first_name, last_name, age, phone_number), product info (set_id, number, theme, subtheme, year, name, pieces, us_price, image_url), and purchase info (quantity). The data has 14 variables and 620 observations. 

In [24]:
#Describe what you see:
print(DF.columns) #lists all column names
print(DF.shape) #tells how big your dataset is
print(DF.dtypes) #tells what kind of data each column holds
DF.describe() #gives a statistical summary of all numeric columns

Index(['first_name', 'last_name', 'age', 'phone_number', 'set_id', 'number',
       'theme', 'subtheme', 'year', 'name', 'pieces', 'us_price', 'image_url',
       'quantity'],
      dtype='object')
(620, 14)
first_name       object
last_name        object
age               int64
phone_number     object
set_id            int64
number           object
theme            object
subtheme         object
year              int64
name             object
pieces          float64
us_price        float64
image_url        object
quantity          int64
dtype: object


age        set_id    year       pieces    us_price    quantity
count  620.000000    620.000000   620.0   551.000000  620.000000  620.000000
mean    34.356452  25124.982258  2018.0   254.206897   29.041613    1.437097
std     11.276537    506.760720     0.0   357.738804   34.630623    0.712849
min     16.000000  24548.000000  2018.0    13.000000    3.990000    1.000000
25%     25.000000  24724.750000  2018.0    70.000000    9.990000    1.000000
50%     33.000000  24804.500000  2018.0   114.000000   19.990000    1.000000
75%     41.000000  25640.250000  2018.0   313.000000   29.990000    2.000000
max     68.000000  26060.000000  2018.0  4634.000000  349.990000    5.000000

In [25]:
# 2 What are the three most common first names of purchasers?
DF['first_name'].value_counts()

first_name
Jackson       13
Joseph        11
Jacob         11
Michael       10
Kaitlyn        8
              ..
Gunnar         1
David          1
Lawrence       1
John'Dunae     1
Kimberly       1
Name: count, Length: 211, dtype: int64

we can see that the three most common first names are Jackson, Joseph, and Jacob.

In [26]:
# 3 What are the three most common themes of Lego sets purchased?
DF['theme'].value_counts()

theme
Star Wars                  75
Nexo Knights               64
Gear                       55
Mixels                     55
City                       45
Friends                    42
Ninjago                    38
Duplo                      35
Bionicle                   34
Creator                    25
DC Comics Super Heroes     22
Elves                      22
Marvel Super Heroes        19
Dimensions                 18
Disney Princess            15
The Angry Birds Movie      11
Technic                    10
Architecture               10
Minecraft                   9
Advanced Models             4
Ghostbusters                3
Seasonal                    3
Collectable Minifigures     3
Ideas                       2
Classic                     1
Name: count, dtype: int64

In this sample, the three most common Lego themes purchased are Star Wars, Nexo Knights, and tied for third is Gear and Mixels.

In [27]:
#4 Among the most common theme of Lego sets purchased, what is the most common subtheme?
common_theme="Star Wars"
mask=DF['theme']==common_theme
DF_common = DF[mask].copy()
DF_common['subtheme'].value_counts()


subtheme
The Force Awakens            15
Buildable Figures            11
Episode V                    10
MicroFighters                10
Battlefront                   7
Original Content              7
Episode III                   6
Rebels                        3
Seasonal                      3
Episode IV                    2
Ultimate Collector Series     1
Name: count, dtype: int64

In this sample, the most common subtheme is The Force Awakens.

In [28]:
#from professor's notebook

#5 Create data frames for each of the ages in the following categories: 
# "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks

# create 5 masks
mask_u18 = (DF['age'] <= 18)
mask_19_25 = (DF['age'] >= 19) & (DF['age'] <=25)
mask_26_35 = (DF['age'] >= 26) & (DF['age'] <=35)
mask_36_50 = (DF['age'] >= 36) & (DF['age'] <=50)
mask_50u = (DF['age'] >= 50)

# create 5 data frames
DF_u18 = DF[mask_u18].copy()
DF_19_25 = DF[mask_19_25].copy()
DF_26_35 = DF[mask_26_35].copy()
DF_36_50 = DF[mask_36_50].copy()
DF_50u = DF[mask_50u].copy()



In [29]:
#6 Which age group has purchased the highest number of Lego sets.

print('under 18:')
print(DF_u18['quantity'].sum())
print('-----')

print('19-25')
print(DF_19_25['quantity'].sum())
print('-----')

print('26-35')
print(DF_26_35['quantity'].sum())
print('-----')

print('36-50')
print(DF_36_50['quantity'].sum())
print('-----')

print('over 50')
print(DF_50u['quantity'].sum())


under 18:
45
-----
19-25
174
-----
26-35
267
-----
36-50
313
-----
over 50
112


In this sample, the age group 36 - 50 spent the most money on Legos.

In [30]:
#from professor's notebook

#7 Which age group has spent the most money on Legos?

# Add a column for money spent.
DF['money_spent'] = DF['us_price']*DF['quantity']

# Then redo the masks
DF_u18 = DF[mask_u18].copy()
DF_19_25 = DF[mask_19_25].copy()
DF_26_35 = DF[mask_26_35].copy()
DF_36_50 = DF[mask_36_50].copy()
DF_50u = DF[mask_50u].copy()

# Then sum the money spent
print('under 18:')
print(DF_u18['money_spent'].sum())
print('-----')

print('19-25')
print(DF_19_25['money_spent'].sum())
print('-----')

print('26-35')
print(DF_26_35['money_spent'].sum())
print('-----')

print('36-50')
print(DF_36_50['money_spent'].sum())
print('-----')

print('over 50')
print(DF_50u['money_spent'].sum())

under 18:
948.5500000000001
-----
19-25
4939.26
-----
26-35
7576.33
-----
36-50
9532.870000000003
-----
over 50
3468.8799999999997


In this sample, the age group 36 - 50 purchased the highest number of Lego sets.

In [31]:
#8 Which Lego theme has made the most money for Lego?
colummns = ['theme','money_spent']
DF[colummns].groupby('theme').sum().sort_values('money_spent',ascending=False)

money_spent
theme                               
Star Wars                    4447.83
Ninjago                      2279.41
City                         2211.35
Nexo Knights                 2209.07
Minecraft                    1549.90
Gear                         1533.21
Friends                      1279.44
Duplo                        1219.50
Elves                        1119.72
Ghostbusters                  879.95
Advanced Models               849.95
Marvel Super Heroes           839.70
Bionicle                      779.54
DC Comics Super Heroes        653.65
Architecture                  589.85
Technic                       572.88
Creator                       549.65
Dimensions                    484.76
The Angry Birds Movie         478.87
Mixels                        414.17
Disney Princess               237.80
Ideas                         209.97
Seasonal                       39.96
Classic                        29.99
Collectable Minifigures        11.97

The Star Wars theme has made the most money for lego in this data.

In [32]:
#9 Which area code has spent the most money on Legos? 
# In the US the area code is the first 3 digits of a phone number.

DF['area_code'] = DF['phone_number'].apply(lambda x: str(x).split('-')[0])

columns = ['area_code','money_spent']
DF[columns].groupby('area_code').sum().sort_values('money_spent',ascending=False)

money_spent
area_code             
nan            3992.64
956             719.96
973             684.91
567             549.86
281             464.90
...                ...
216               9.99
509               9.99
308               9.99
325               9.98
636               4.99

[157 rows x 1 columns]

In this sample, looks like unknown regions purchased the most but the area code 956 has spent the most money on Legos.

In [33]:
counts = DF['area_code'].value_counts()
counts

area_code
nan    92
281     9
441     9
567     9
712     9
       ..
760     1
563     1
917     1
947     1
402     1
Name: count, Length: 157, dtype: int64

In [40]:
spent = DF[columns].groupby('area_code').sum().sort_values('money_spent',ascending=False)

average_spent = []
for c in counts.items():
    average_spent.append(spent.loc[c[0]]['money_spent']/c[1])
    # if c[0] != 'nan':
    #     print(f'Area code {c[0]} spent an average of {spent.loc[c[0]]['money_spent']/c[1]}')

spent['average_spent'] = average_spent
spent.sort_values('average_spent',ascending=False)

money_spent  average_spent
area_code                            
647              79.95     359.980000
330              39.98     169.980000
787              49.99     144.990000
662              32.97     124.990000
856              79.96     114.985000
...                ...            ...
313              94.95       8.323333
862              74.97       7.490000
617              59.97       5.490000
413              19.98       4.990000
409              65.96       4.990000

[157 rows x 2 columns]

First we calculate how much money each purchase is worth (us_price * quantity).

Then we group by area code and total up the money.

We count how many purchases came from each area code.

Using a for loop, we divide the total money by the number of purchases → this gives us the average spent per purchase in that area code.

Finally, we sort the table so the area code with the highest average spending shows up at the top.